packages/legend-application-query/src/stores/QueryBuilderChangeDetectionState.ts

```diff
@@ -0,0 +1,161 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { RawLambda } from '@finos/legend-graph';
+import {
+  type GeneratorFn,
+  hashObject,
+  LogEvent,
+  promisify,
+  assertErrorThrown,
+  IllegalStateError,
+} from '@finos/legend-shared';
+import {
+  action,
+  flow,
+  type IReactionDisposer,
+  makeObservable,
+  observable,
+  isComputedProp,
+  reaction,
+} from 'mobx';
+import type { QueryBuilderState } from './QueryBuilderState.js';
+import { keepAlive, type IDisposer } from 'mobx-utils';
+import { QUERY_BUILDER_CHANGE_DETECTION_EVENT } from '../QueryBuilder_Const.js';
+
+export class QueryBuilderChangeDetectionState {
+  querybuilderState: QueryBuilderState;
+  queryHashCode = hashObject(new RawLambda(undefined, undefined));
+  isEnabled = false;
+  initialQueryHashCode?: string | undefined;
+  currentQueryHashCode?: string | undefined;
+  stateHashIndex: string | undefined;
+  isBuildingStateHashesIndex = false;
+  /**
+   * Keep the list of disposers to deactivate `keepAlive` for computed value of element hash code.
+   * See {@link preComputeGraphElementHashes} for more details
+   */
+  private stateHashCodeKeepAliveComputationDisposers: IDisposer[] = [];
+  private changeDetectionReaction?: IReactionDisposer | undefined;
+
+  constructor(queryBuilderState: QueryBuilderState) {
+    makeObservable(this, {
+      initialQueryHashCode: observable,
+      currentQueryHashCode: observable,
+      stateHashIndex: observable,
+      isBuildingStateHashesIndex: observable,
+      setInitialQueryHashCode: action,
+      setCurrentQueryHashCode: action,
+      detectChanges: flow,
+      setIsBuildingStateHashesIndex: action,
+      buildStateHashesIndex: flow,
+    });
+
+    this.querybuilderState = queryBuilderState;
+  }
+
+  setQueryHashCode(val: string): void {
+    this.queryHashCode = val;
+  }
+
+  setIsEnabled(val: boolean): void {
+    this.isEnabled = val;
+  }
+
+  setInitialQueryHashCode(val: string | undefined): void {
+    this.initialQueryHashCode = val;
+  }
+
+  setCurrentQueryHashCode(val: string | undefined): void {
+    this.currentQueryHashCode = val;
+  }
+
+  setIsBuildingStateHashesIndex(val: boolean): void {
+    this.isBuildingStateHashesIndex = val;
+  }
+
+  async buildHashesIndex(): Promise<string> {
+    return promisify(() => this.querybuilderState.hashCode);
+  }
+
+  *detectChanges(): GeneratorFn<void> {
+    //TODO: Set states to indicate that query has got changed and query results have gone stale
+  }
+
+  *buildStateHashesIndex(): GeneratorFn<void> {
```

Comment:
--> I think we can merge `preCompute` and this method into one since you call them together every time anyway. In fact, you can modify the current `start()` method that will, in sequence:
1. pre-compute the hash
2. set the initial hash
3. start() change detection reaction

and all them in all other places